### PR TITLE
Fix null pointer problems on windows.

### PR DESCRIPTION
**Pull request by joe2018Outlookcom**
Annotation 2019-05-14 112445

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
Any reason for the serial open / close sequence?

File: crypto/threads_win.c [hunk 1]
```diff
@@ -39,18 +39,27 @@ CRYPTO_RWLOCK *CRYPTO_THREAD_lock_new(void)
 
 int CRYPTO_THREAD_read_lock(CRYPTO_RWLOCK *lock)
 {
+    if (NULL == lock)
+        return 0;
```
Comment (mspncp) [inline crypto/threads_win.c]: Looking at the callstack in the screenshot, it is clear that it's the `rand_meth_lock` which is uninitialized. However, this  lock should be initialized at first call via `do_rand_init()` and then be valid until ` rand_cleanup_int()` is called at program exit. If this pointer is NULL then something has fundamentally gone wrong and the reason needs to be found and fixed instead.

https://github.com/openssl/openssl/blob/e9f148c9356b18995298f37bafbf1836a3fce078/crypto/rand/rand_lib.c#L726-L733

**Comment by mspncp** [comment]
@joe2018Outlookcom if you are encountering this crash regularly, please open an issue (or reopen this pull request) and provide more information about the steps to reproduce the crash, together with the output of the command `perl configdata.pm --dump`.

**Comment by joe2018Outlookcom** [comment]
Reopen for cla-check.

**Comment by joe2018Outlookcom** [comment]
This PR looks like a bad submission. However, it must be noted that **crashes occur when a process exits** while **no problems occur before it exits**.

Have noticed that([http://openssl.6102.n7.nabble.com/1-1-1b-crash-RUN-ONCE-problem-td76153.html](http://openssl.6102.n7.nabble.com/1-1-1b-crash-RUN-ONCE-problem-td76153.html))：
------------------------------------
> main() must not exit while threads are running code from 
> libraries with exit handlers.  This is issue is not unique to 
> OpenSSL.  I've observed the same problem with threading running 
> code in MIT's GSSAPI library. 
------------------------------------
It seems that null-pointer problem occurs here for avoiding any memory leaks while the library de-initialized(when exit）. When the library has been de-initialized (note that there only left the main thread when a problem occurs), the critical operation here (even if a valid pointer is passed in) is meaningless.

PS: EnterCriticalSection requires passing in a valid CITICAL_SECTION pointer ([https://docs.microsoft.com/en-us/windows/desktop/Sync/using-critical-section-objects](https://docs.microsoft.com/en-us/windows/desktop/Sync/using-critical-section-objects)), so when passing in a null pointer, the problem will naturally arise; so using this Win32 API(just follow the contracts)  with null pointer protection should be a code style (see [https://github.com/openssl/openssl/blob/da89ac0b25c5198bb0d707d7545bea7bc1c5a01e/crypto/threads_win.c#L58-L64](https://github.com/openssl/openssl/blob/da89ac0b25c5198bb0d707d7545bea7bc1c5a01e/crypto/threads_win.c#L58-L64) , the same null pointer protection) or good coding habits.

PS2: perl configdata.pm –dump  can certainly provide, if necessary, but it must be noted that it may not be necessary to provide overly redundant information to interfere with the thinking of solving the problem (just using static CRT under Win32 without removing multithreading support). At the same time, demo code that can reproduce the problem 100% is really very effective, but it's difficult to provide. Another thing to add is that simply replacing OpenSSL 1.0.2r will eliminate the problem.

**Comment by slontis** [comment]
There must be an ordering issue with your shutdown. Adding the NULL checks may just hide issues.
Try to figure out when rand_cleanup_int() gets called. (Which sets the rand_meth_lock to NULL).
If OPENSSL_cleanup() is getting called early that would be bad..
If its not and the rand_meth_lock is set to NULL through some other means, then try to figure out how/when that happens.

**Comment by paulidale** [comment]
Closing this, I can't see the changes being merged.

**Pull request closed without merging**